test/python/test_qobj.py

```diff
@@ -39,7 +40,82 @@ def test_create_qobj(self):
                     {'name': 'u1', 'params': [0.4], 'qubits': [1]},
                     {'name': 'u2', 'params': [0.4, 0.2], 'qubits': [1]}
                 ]}
-            ]
+            ],
         }
 
-        self.assertEqual(qobj.as_dict(), expected)
+    def tearDown(self):
+        self.valid_qobj = None
+        self.expected_dict = None
+
+    def test_create_qobj_and_req_args(self):
+        """Test creation of a Qobj and check for its required args."""
+        self.assertTrue(
+            all(getattr(self.valid_qobj, required_arg) is not None
+                for required_arg in Qobj.REQUIRED_ARGS))
+
+        with self.assertRaises(ValueError):
+            Qobj(qobj_id=None, header=None, config=None, experiments=None)
+
+    def test_as_dict(self):
+        """Test conversion to dict of a Qobj based on the individual elements."""
+        self.valid_qobj._version = '67890'  # private member variables shouldn't appear in the dict
+        self.assertEqual(self.valid_qobj.as_dict(), self.expected_dict)
+
+    def test_as_dict_to_json(self):
+        """Test dictionary representation of Qobj against its schema."""
+        file_path = self._get_resource_path('qobj_schema.json', qiskit_path.SCHEMAS)
+
+        with open(file_path, 'r') as schema_file:
+            schema = json.load(schema_file)
+
+        jsch.validate(self.valid_qobj.as_dict(), schema)
+
+
+class TestQobjConfig(QiskitTestCase):
+    """Tests for QobjConfig."""
+    def test_init_qobj_config(self):
+        """Test initialization of a QobjConfig for required arguments."""
+        shots = 1
+        memory_slots = 2
+
+        qobj_config = QobjConfig(shots=shots, memory_slots=memory_slots)
+
+        self.assertTrue(
+            all(getattr(qobj_config, required_arg) is not None
+                for required_arg in QobjConfig.REQUIRED_ARGS))
+
+        with self.assertRaises(ValueError):
```

Comment:
It is not implemented for a `Qobj` to raise in case there is a required argument missing. As commented offline, talk to @diego-plan9 about adding this functionality to the class. What is sure is that `from_dict` behaves in a very similar way as your test expect. Add tests for `from_dict` based on this one.

Reply:
@diego-plan9 and I decided that TypeErrors should be raised if the required arguments are not provided, as Python does by default. The assertions that the errors are raised helps validate that the errors haven't been circumvented somehow, such as being intercepted by a try/except block. ```from_dict``` is in the process of being removed so I won't add tests based on this one.

Reply:
Elaborating a bit - I basically see arguments for any of the three options:
* removing the `with self.assertRaises` block - technically, we are testing Python functionality (ie. a function, in this case the constructor, needs the positional parameters), which is a given.
* keeping the `with self.assertRaises` block - with this, we are making explicit that we satisfy the original requirement ("a QobjFoo should have bar, baz as required attributes"), decoupled from the current implementation.
* changing the test so it uses `from_dict`  instead of instantation: `from_dict`, even if it is going to be hopefully short-lived, is also responsible from checking that  the attributes are present.

So basically whichever @delapuente and @Max-Reuter agree upon sounds good to me!

Reply:
I did not realize mandatory arguments are modelled as positional arguments without defaults so there is no need for extra validation. My only concern is passing `None` which would trick Python into generating an invalid dictionary representation. But let's move that effort to #654.

Although it's true that `from_dict` is on its way of being removed, the functionality won't disappear (see #652) so these test will remain valid.

So, I would recommend creating tests for `from_dict` and testing that validation works there. @Max-Reuter what do you think?

Reply:
I propose we remove the ```assertRaises``` block altogether and rely on lines 84-85 to handle the ```None``` case, if we don’t need to check that the error is being raised and not circumvented. I don’t see the use of ```from_dict``` if all we’re doing is checking the arguments aren’t ```None```. Python will handle the case where the arguments aren’t passed at all.